### PR TITLE
Implemented : PDFs optimization

### PR DESCRIPTION
**Pull request by a5i**
Implemented PDF size optimizations:
 - UseCrossReferenceStream (7.5.8 Cross-Reference Streams)
 - UseObjectStreams (7.5.7  Object Streams)
 - CombineDuplicateDirectObjects
 - CombineDuplicateStreams
 - ImageQuality

ref #181

### CODE REVIEW

**Comment by a5i** [comment]
- Added Optimizer interface
- Added ImageUpperPPI optimization
- Added combine identical indirect objects optimization

File: pdf/model/optimize/image_ppi.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+)
+
+// ImagePPI optimizes images.
+// It implements interface model.Optimizer.
+// Does not support inline images yet.
+type ImagePPI struct {
+	ImageUpperPPI float64
+}
+
+func scaleImage(stream *PdfObjectStream, scale float64) error {
```
Comment (gunnsth) [inline pdf/model/optimize/image_ppi.go]: When handling the error up from scaleImage, would be better to output a Debug ERROR message and leave object unmodified.

File: pdf/model/optimize/image_ppi.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	"image"
+	"math"
+
+	"github.com/unidoc/unidoc/pdf/contentstream"
+	. "github.com/unidoc/unidoc/pdf/core"
+	"github.com/unidoc/unidoc/pdf/model"
+	"golang.org/x/image/draw"
+)
+
+// ImagePPI optimizes images.
```
Comment (gunnsth) [inline pdf/model/optimize/image_ppi.go]: Maybe make the comment a bit more descriptive. E.g.
ImagePPI optimizes images by scaling images such that the PPI (pixels per inch) is never higher than `ImageUpperPPI`.

And // TODO: Add support for inline images.

File: pdf/model/optimize/optimizer.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+package optimize
+
+import (
+	. "github.com/unidoc/unidoc/pdf/core"
+	"github.com/unidoc/unidoc/pdf/model"
+)
+
+// New creates a optimizers chain from options
```
Comment (gunnsth) [inline pdf/model/optimize/optimizer.go]: Usually recommended to return the implementation (Chain).
Any disadvantage of that here?

File: pdf/model/writer.go [hunk 4]
```diff
@@ -10,21 +10,36 @@ package model
...
 	"crypto/rand"
+	"encoding/binary"
 	"errors"
 	"fmt"
 	"io"
 	"math"
 	"strings"
 	"time"
 
+	_ "image/jpeg"
```
Comment (gunnsth) [inline pdf/model/writer.go]: Are these still needed after refactoring to optimize package?

File: pdf/model/writer.go [hunk 5]
```diff
@@ -132,6 +150,131 @@ func NewPdfWriter() PdfWriter {
...
+		common.Log.Info("TODO(a5i): implement copyObject for %+v", obj)
+	}
+	// return other objects as is
+	return obj
+}
+
+// copyObjects makes objects copy and set as working
+func (this *PdfWriter) copyObjects(replaceTable map[PdfObject]PdfObject) {
+	if replaceTable == nil {
+		replaceTable = make(map[PdfObject]PdfObject)
```
Comment (gunnsth) [inline pdf/model/writer.go]: Can you briefly describe the replaceTable? Seems to me like used as an endpoint for recursion, is that correct?

File: pdf/model/writer.go [hunk 6]
```diff
@@ -152,6 +295,16 @@ func (this *PdfWriter) SetOCProperties(ocProperties PdfObject) error {
 	return nil
 }
 
+// SetOptimizer sets the optimizer to optimize PDF before writing
```
Comment (gunnsth) [inline pdf/model/writer.go]: dot

File: pdf/model/writer.go [hunk 7]
```diff
@@ -152,6 +295,16 @@ func (this *PdfWriter) SetOCProperties(ocProperties PdfObject) error {
...
+func (this *PdfWriter) SetOptimizer(optimizer Optimizer) {
+	this.optimizer = optimizer
+}
+
+// GetOptimizer returns current PDF optimizer
```
Comment (gunnsth) [inline pdf/model/writer.go]: dot

File: pdf/model/optimize/combine_duplicate_direct_objects.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	for _, obj := range objects {
+		ind, isIndirectObj := obj.(*PdfIndirectObject)
+		if !isIndirectObj {
+			continue
+		}
+		if dict, isDictObj := ind.PdfObject.(*PdfObjectDictionary); isDictObj {
+			if tp, ok := GetName(dict.Get(PdfObjectName("Type"))); ok && *tp == "Page" {
+				processDict(dict.Get(PdfObjectName("Resources")).(*PdfObjectDictionary))
```
[the remark this answers is not in the document]
Comment (a5i) [reply pdf/model/optimize/combine_duplicate_direct_objects.go]: I think about performance optimization. It should work for all direct objects. I will change the code.

File: pdf/model/writer.go (hunk 5 above)
Comment (a5i) [reply pdf/model/writer.go]: Yes, You are right it's endpoint for recursion.
I renamed  replaceTable to objectToObjectCopyMap and added a description.

**Pull request merged**